endpoints/openrtb2/auction.go

```diff
@@ -358,6 +367,20 @@ func (deps *endpointDeps) validateRequest(req *openrtb.BidRequest) []error {
 	return errL
 }
 
+func validateAndFillSourceTID(req *openrtb.BidRequest) error {
+	if req.Source == nil {
+		req.Source = &openrtb.Source{}
+	}
+	if req.Source.TID == "" {
+		if rawUUID, err := uuid.NewV4(); err == nil {
+			req.Source.TID = rawUUID.String()
+		} else {
+			return errors.New("req.Source.TID missing in the req and error creating a random UID")
+		}
+	}
+	return nil
+}
```

Comment:
I believe latter on the execution flow in function `prepareSource`, the `Source` field might get overwritten (line 216 below). Should we prevent this and also add a test that runs either `prepareSource` or even other function lower in the stack to make sure that a given bid request with a random `Source.TID` doesn't get overwritten?
```
190 func prepareSource(req *openrtb.BidRequest, bidder string, sChainsByBidder map[string]*openrtb_ext.ExtRequestPrebidSChainSChain) {
191     const sChainWildCard = "*"
192     var selectedSChain *openrtb_ext.ExtRequestPrebidSChainSChain
193
194     wildCardSChain := sChainsByBidder[sChainWildCard]
195     bidderSChain := sChainsByBidder[bidder]
196
197     // source should not be modified
198     if bidderSChain == nil && wildCardSChain == nil {
199         return
200     }
201
202     if bidderSChain != nil {
203         selectedSChain = bidderSChain
204     } else {
205         selectedSChain = wildCardSChain
206     }
207
208     // set source
209     var source openrtb.Source
210     schain := openrtb_ext.ExtRequestPrebidSChain{
211         SChain: *selectedSChain,
212     }
213     sourceExt, err := json.Marshal(schain)
214     if err == nil {
215         source.Ext = sourceExt
216         req.Source = &source //<-- gets written
217     }
218 }
exchange/utils.go
```

Reply:
I believe this is more of a bug in the `prepareSource` function. We should be checking if a field exists in the request before replacing it. I added that check to this function so that it is creating a new `req.Source` object only if it wasn't present in the request to start with. This way the `TID` value or any other values in the `req.Source` object won't get overridden. 

That was a great catch though! Thanks :)